Optimization in asSeenFrom.

paulp opened the pull request:
Eliminated about half the allocations of AsSeenFromMap by examining the
arguments more closely. The impact is not what one might conclude from
that statistic, because those were the cheap asSeenFrom calls - but now
they are cheaper. Also,
- worked in some eager vals rather than lazy ones.
- changed hot-spot comparison to use eq
- simplified annotationArgRewriter
- elimnated separate tracking of capturedPre

Statistics from compiling scala/collection/**.scala:
- 4209382 calls to asSeenFrom
- 1025945 maps created (instantiations of AsSeenFromMap)
- 1525649 maps avoided (these were instantiations before this patch)

Review by most honorable performance-san @gkossakowski .

paulp committed:
Optimization in asSeenFrom.

Eliminated about half the allocations of AsSeenFromMap by examining the
arguments more closely. The impact is not what one might conclude from
that statistic, because those were the cheap asSeenFrom calls - but now
they are cheaper. Also,

 - worked in some eager vals rather than lazy ones.
 - changed hot-spot comparison to use eq
 - simplified annotationArgRewriter
 - elimnated separate tracking of capturedPre

Statistics from compiling scala/collection/**.scala:

  - 4209382 calls to asSeenFrom
  - 1025945 maps created (instantiations of AsSeenFromMap)
  - 1525649 maps avoided (these were instantiations before this patch)

Review by most honorable performance-san @gkossakowski .

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/435/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/435/

adriaanm commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -695,17 +695,25 @@ trait Types extends api.Types { self: SymbolTable =>
+        val pre1 = pre.normalize
 
+        val result: Type = (
+          if (pre1.isTrivial && (clazz.isPackageClass || !clazz.isClass)) this
```
can this condition be factored out and shared with AsSeenFromMap?

adriaanm commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -4254,44 +4264,41 @@ trait Types extends api.Types { self: SymbolTable =>
-          qvar
-      }).tpe
+    def stabilize(pre: Type, clazz: Symbol): Type = {
+      capturedParams find (_.owner == clazz) match {
```
This refactoring is not obvious to me. Could you elucidate, por favor?

paulp replied:
Formerly "capturedPre" was maintained as a map of Symbol -> TypeParam.  If clazz -> qvar is already in the map, then stabilize returns qvar.tpe.  Otherwise it updates the map with clazz = new existential.

So what that means is: the list of captured params uniquely identifies the last of captured prefixes, because they are the owners of those params.  If it would have been in the map, it will be the owner of a member of capturedParams, and if it would not have been in the map, then it won't (this being assured by the "getOrElseUpdate" logic being applied - every call to stabilize either finds a pre-existing qvar for a clazz or creates one, so there can't be two.)

That is perhaps not my clearest explanation ever, let me know.

paulp replied:
Oh, and for very small sets like these, looking through a list is way faster than managing an immutable map.

adriaanm commented:
Except for the lines on which I commented, the refactoring improves understandability of the code as far as I'm concerned.

paulp commented:
I collected a bunch of numbers but I didn't know what to do with that code.  That's a general problem it would be nice to do something about - I build up and take down scaffolding far too often.  Maybe we can use macros somewhere to alleviate this.  As I said, about half the instantiations of AsSeenFromMap were eliminated, and I posted a statistic from a subset of the compiler.

You may need to constrain your memory in order to see the improvement, because we're mostly eliminating the creation of garbage, and if there's never any pressure on the collector you may never feel it.  My time measurements were small but reproducible improvements; the profiler output was far more convincing, because it's simply doing less work.

paulp commented:
I collected more numbers. It was more of a slam dunk than I thought.
Over 20% reduction in time spent in asSeenFrom. I ran the following
command line was run three times on each build, with the only difference
this commit. Each compiler was built optimized.

```
scalac -J-verbose:gc -J-Xmx384m -Ystatistics -d /tmp \
  $(find /scala/trunk/src/library/scala/collection -name '*.scala')
```

The asSeenFrom statistics:

```
asseen.txt:time spent in asSeenFrom      : 608996 spans, 3845401ms (18.7%) aggregate, 2977224ms (14.5%) specific
asseen.txt:time spent in asSeenFrom      : 608996 spans, 3896184ms (18.3%) aggregate, 3055034ms (14.3%) specific
asseen.txt:time spent in asSeenFrom      : 608996 spans, 3886309ms (18.5%) aggregate, 3043777ms (14.5%) specific

master.txt:time spent in asSeenFrom      : 608882 spans, 4484066ms (20.5%) aggregate, 3804826ms (17.4%) specific
master.txt:time spent in asSeenFrom      : 608882 spans, 4569611ms (20.9%) aggregate, 3871168ms (17.7%) specific
master.txt:time spent in asSeenFrom      : 608882 spans, 4463335ms (21.0%) aggregate, 3777003ms (17.8%) specific
```

Wall clock is less obvious, but with numbers like the above it seems irrelevant.
I don't have a world-class test isolation rig so there's going to be jitter.

```
asseen.times:real   1m10.471s
asseen.times:real   1m10.755s
asseen.times:real   1m11.694s

master.times:real   1m12.818s
master.times:real   1m13.418s
master.times:real   1m11.944s
```

Total garbage collections across all three runs:

```
master    1235
asseen    1221
```

paulp committed:
Further optimization of asSeenFrom.

After some more early exclusion of uninteresting arguments,
asSeenFrom is now over twice as fast.  Compiling the collections,
three runs:

  -- Without these two commits:
  8108775ms, 8132052ms, 8257527ms in asSeenFrom
  Wall clock: 0m54.202s, 0m54.877s, 0m55.925s

  -- With these two commits:
  3568719ms, 3542860ms, 3570892ms in asSeenFrom
  Wall clock: 0m50.730s, 0m51.213s, 0m50.850s

paulp commented:
Adriaan's comment about pulling that snippet out made me notice I wasn't done yet.  With another commit (not attached as I haven't run tests yet) the numbers arrive here.  To ensure it was a fair fight, I modified asSeenFrom in master and my branch to count the entire time spent in asSeenFrom, not just the subset which reach the map as is done in master now.

```
time scalac -Ystop-after:typer -J-Xmx512m -Ystatistics \
  $(find /scratch/trunk1/src/library/scala/collection -name '*.scala')
```

master:

```
time spent in asSeenFrom      : 4100067 spans, 13828772ms (28.1%) aggregate, 8108775ms (16.5%) specific
time spent in asSeenFrom      : 4100069 spans, 11629263ms (23.4%) aggregate, 8132052ms (16.4%) specific
time spent in asSeenFrom      : 4100065 spans, 14850108ms (29.3%) aggregate, 8257527ms (16.3%) specific
real    0m54.202s
real    0m54.877s
real    0m55.925s
```

me:

```
time spent in asSeenFrom      : 4076900 spans, 7158291ms (15.7%) aggregate, 3568719ms (7.8%) specific
time spent in asSeenFrom      : 4076916 spans, 9323734ms (20.2%) aggregate, 3542860ms (7.7%) specific
time spent in asSeenFrom      : 4076908 spans, 9255200ms (20.2%) aggregate, 3570892ms (7.8%) specific
real    0m50.730s
real    0m51.213s
real    0m50.850s
```

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/448/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/448/

gkossakowski commented:
Paul, I'm in the process of creating a tool that will complement what -Ystatistics. I hope to get really basic version done in a few days. This pull request would be an excellent test-case for it. Can it wait?

Also, for commits like this we usually need to reviewers: one making sure that semantics are not being broken (@adriaanm jumped on this even if we didn't ask) and the other one making sure that performance is indeed being improved. Especially when it comes to typer I need a help on semantics side for now.

paulp commented:
I'd rather not wait - the evidence that this patch is benefit is pretty overwhelming from where I sit, and leaving it open means one more thing for me to keep track of, one more piece of non-master code I have to assume in my other branches.  You can test the tool on it as easily if has been merged as you can if it's still sitting here.

gkossakowski commented:
Fair enough. I'll have a look at it either today or on Monday.

gkossakowski commented:
Ok, sorry for the delay. I benchmarked this independently and I got really nice results! This pull request gives 8%-10% speedup on all of my tests. To understand why see comparison method counts before this change was introduced and with this change:

```
                                            Name                                                      |  Time Diff (ms)  |  Old Time (ms)  |  New Time (ms)  |  Inv. Count Diff  |  Old Inv. Count  |  New Inv. Count  |
scala.reflect.internal.SymbolTable.NoType()                                                           |             -58  |             88  |             29  |         -766,045  |       1,151,631  |         385,586  |
                                                                                                      |                  |                 |                 |                   |                  |                  |
java.lang.Object.equals(Object)                                                                       |             -69  |            398  |            329  |         -620,385  |       4,836,684  |       4,216,299  |
                                                                                                      |                  |                 |                 |                   |                  |                  |
scala.reflect.internal.SymbolTable.NoPrefix()                                                         |             -37  |             81  |             43  |         -485,245  |       1,055,521  |         570,276  |
                                                                                                      |                  |                 |                 |                   |                  |                  |
scala.reflect.internal.SymbolTable.phaseId(int)                                                       |             -53  |            332  |            278  |         -459,645  |       4,311,324  |       3,851,679  |
                                                                                                      |                  |                 |                 |                   |                  |                  |
scala.reflect.internal.Symbols$ClassSymbol.isClass()                                                  |             -39  |             83  |             43  |         -456,316  |       1,030,859  |         574,543  |
                                                                                                      |                  |                 |                 |                   |                  |                  |
scala.reflect.internal.util.Statistics$.incCounter(Statistics$Counter)                                |             -42  |            207  |            164  |         -413,403  |       2,443,498  |       2,030,095  |
                                                                                                      |                  |                 |                 |                   |                  |                  |
scala.reflect.internal.Types$SimpleTypeProxy$class.baseType(Types$SimpleTypeProxy, Symbols$Symbol)    |          -4,526  |         12,771  |          8,244  |         -383,598  |         409,673  |          26,075  |
                                                                                                      |                  |                 |                 |                   |                  |                  |
scala.reflect.internal.Types$SingletonType.baseType(Symbols$Symbol)                                   |          -4,529  |         12,780  |          8,250  |         -383,598  |         409,673  |          26,075  |
                                                                                                      |                  |                 |                 |                   |                  |                  |
scala.reflect.internal.Types$SubstMap$$anonfun$13.<init>(Types$SubstMap)                              |            -221  |            221  |              0  |         -327,396  |         327,396  |               0  |
                                                                                                      |                  |                 |                 |                   |                  |                  |
scala.reflect.internal.Types$ThisType.underlying()                                                    |            -181  |            699  |            518  |         -267,700  |         413,185  |         145,485  |
                                                                                                      |                  |                 |                 |                   |                  |                  |
scala.Predef$.assert(boolean)                                                                         |             -30  |            168  |            137  |         -246,895  |       2,113,696  |       1,866,801  |
                                                                                                      |                  |                 |                 |                   |                  |                  |
scala.reflect.internal.Types$TypeRef.equals(Object)                                                   |              -2  |            591  |            588  |         -240,289  |         403,875  |         163,586  |
                                                                                                      |                  |                 |                 |                   |                  |                  |
scala.reflect.internal.Symbols$Symbol.owner()                                                         |             -79  |            281  |            201  |         -227,163  |         594,657  |         367,494  |
                                                                                                      |                  |                 |                 |                   |                  |                  |
scala.reflect.internal.Types$AsSeenFromMap.base(Types$Type, Symbols$Symbol)                           |          -3,604  |          8,992  |          5,387  |         -221,163  |         238,956  |          17,793  |
                                                                                                      |                  |                 |                 |                   |                  |                  |
scala.reflect.internal.Types$ClassTypeRef$class.baseType(Types$ClassTypeRef, Symbols$Symbol)          |          -4,209  |         17,409  |         13,200  |         -221,163  |         259,792  |          38,629  |
                                                                                                      |                  |                 |                 |                   |                  |                  |
scala.reflect.internal.Types$ModuleTypeRef.baseType(Symbols$Symbol)                                   |            -111  |            376  |            264  |         -221,163  |         222,687  |           1,524  |
                                                                                                      |                  |                 |                 |                   |                  |                  |
scala.reflect.internal.Types$TypeRef.prefix()                                                         |             -15  |             17  |              1  |         -221,163  |         240,194  |          19,031  |
                                                                                                      |                  |                 |                 |                   |                  |                  |
scala.reflect.internal.Symbols$ModuleClassSymbol.typeOfThis()                                         |              -5  |            230  |            225  |         -208,878  |         307,961  |          99,083  |
                                                                                                      |                  |                 |                 |                   |                  |                  |
scala.reflect.internal.Types$SingleType.underlying()                                                  |             -10  |             86  |             75  |         -208,878  |         315,513  |         106,635  |
                                                                                                      |                  |                 |                 |                   |                  |                  |
scala.reflect.internal.BaseTypeSeqs$BaseTypeSeq.length()                                              |             -20  |             85  |             64  |         -182,960  |       1,119,815  |         936,855  |
                                                                                                      |                  |                 |                 |                   |                  |                  |
scala.reflect.internal.Types$CompoundType.baseTypeSeq()                                               |          -2,388  |         21,018  |         18,630  |         -182,628  |       1,039,653  |         857,025  |
                                                                                                      |                  |                 |                 |                   |                  |                  |
scala.reflect.internal.Symbols$Symbol.info()                                                          |          -2,782  |         50,792  |         48,009  |         -182,620  |       1,649,955  |       1,467,335  |
                                                                                                      |                  |                 |                 |                   |                  |                  |
scala.reflect.internal.Symbols$Symbol.rawInfo()                                                       |          -2,478  |         35,879  |         33,401  |         -181,941  |       1,864,740  |       1,682,799  |
                                                                                                      |                  |                 |                 |                   |                  |                  |
scala.reflect.internal.Types$AsSeenFromMap.apply(Types$Type)                                          |          -7,007  |         21,892  |         14,885  |         -181,596  |         248,006  |          66,410  |
                                                                                                      |                  |                 |                 |                   |                  |                  |
scala.reflect.internal.Symbols$Symbol.needsFlatClasses()                                              |             -18  |            612  |            594  |         -155,892  |       1,027,005  |         871,113  |
                                                                                                      |                  |                 |                 |                   |                  |                  |
scala.tools.nsc.Global$GlobalPhase.flatClasses()                                                      |             -12  |             82  |             70  |         -155,892  |       1,027,031  |         871,139  |
                                                                                                      |                  |                 |                 |                   |                  |                  |
scala.reflect.internal.Symbols$ClassSymbol.owner()                                                    |            -246  |            896  |            650  |         -153,229  |         441,030  |         287,801  |
                                                                                                      |                  |                 |                 |                   |                  |                  |
scala.reflect.internal.Symbols$TypeSymbol.isNonBottomSubClass(Symbols$Symbol)                         |          -3,117  |         20,087  |         16,970  |         -140,501  |         256,764  |         116,263  |
                                                                                                      |                  |                 |                 |                   |                  |                  |
scala.reflect.internal.Symbols$Symbol.baseTypeSeqLength$1(Symbols$Symbol)                             |          -2,188  |         16,807  |         14,619  |         -138,176  |         839,342  |         701,166  |
                                                                                                      |                  |                 |                 |                   |                  |                  |
scala.reflect.internal.Symbols$TypeSymbol.isType()                                                    |             -19  |             91  |             72  |         -137,815  |       1,159,351  |       1,021,536  |
                                                                                                      |                  |                 |                 |                   |                  |                  |
scala.reflect.internal.Types$TypeMap.mapOver(Types$Type)                                              |          -5,824  |         30,649  |         24,825  |         -135,508  |       1,015,864  |         880,356  |
```

We save a lot of calls here.

gkossakowski commented:
I've got so excited that I forgot to say LGTM. Color me impressed.

adriaanm commented:
Excellent. I'd love to merge it as soon as M5 has been cut (if Martin agrees) but note that it doesn't merge cleanly anymore.

paulp commented:
The merge conflict is from my removal of def base from the map; I can fix it trivially, but I'll wait until it's going into master.

odersky commented:
LGTM from me also. Great work!

paulp committed:
Merge branch 'master' into topic/optimize-asSeenFrom

Conflicts:
	src/reflect/scala/reflect/internal/Types.scala

paulp commented:
I'm not sure what the policy is when the history starts getting merge-y, let me know if I'm supposed to clean up the history.

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/505/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/505/

adriaanm commented:
rebased onto 2.10.x as e90f79a446 and e6ac99e6a9
